aequilibrae/project/project_creation.py

```diff
@@ -0,0 +1,218 @@
+                      direction INTEGER NOT NULL DEFAULT 0,
+                      distance NUMERIC,
+                      modes TEXT NOT NULL,
+                      link_type TEXT NOT NULL DEFAULT 'default',
+                      {}
+                      FOREIGN KEY (link_type)
+                            REFERENCES link_types (link_type));"""
+
+    flds = fields["one-way"]
+
+    # returns first key in the dictionary
+    def fkey(f):
+        return list(f.keys())[0]
+
+    owlf = ["{} {}".format(fkey(f), f[fkey(f)]["type"]) for f in flds if fkey(f).lower() not in req_link_flds]
+
+    flds = fields["two-way"]
+    twlf = []
+    for f in flds:
+        nm = fkey(f)
+        tp = f[nm]["type"]
+        twlf.extend([f"{nm}_ab {tp}", f"{nm}_ba {tp}"])
+
+    link_fields = owlf + twlf
+
+    if link_fields:
+        sql = sql.format(",".join(link_fields) + ",")
+    else:
+        sql = sql.format("")
+
+    curr.execute(sql)
+
+    sql = """CREATE TABLE 'nodes' (ogc_fid INTEGER PRIMARY KEY,
+                             node_id INTEGER UNIQUE NOT NULL,
+                             is_centroid INTEGER NOT NULL DEFAULT 0,
+                             modes VARCHAR,
+                             link_types VARCHAR {});"""
+
+    flds = parameters["network"]["nodes"]["fields"]
+    ndflds = [f"{fkey(f)} {f[fkey(f)]['type']}" for f in flds if fkey(f).lower() not in req_node_flds]
+
+    if ndflds:
+        sql = sql.format("," + ",".join(ndflds))
+    else:
+        sql = sql.format("")
+    curr.execute(sql)
+
+    curr.execute("""SELECT AddGeometryColumn( 'links', 'geometry', 4326, 'LINESTRING', 'XY' )""")
+    curr.execute("""SELECT AddGeometryColumn( 'nodes', 'geometry', 4326, 'POINT', 'XY' )""")
+    conn.commit()
+
+
+def add_triggers(conn) -> None:
+    """Adds consistency triggers to the project"""
+    add_network_triggers(conn)
+    add_mode_triggers(conn)
+    add_link_type_triggers(conn)
+
+
+def add_network_triggers(conn) -> None:
+    logger.info("Adding network triggers")
+    add_trigger_from_file(conn, "network_triggers.sql")
+
+
+def add_mode_triggers(conn) -> None:
+    logger.info("Adding mode table triggers")
+    add_trigger_from_file(conn, "modes_table_triggers.sql")
+
+
+def add_link_type_triggers(conn) -> None:
+    logger.info("Adding link type table triggers")
+    add_trigger_from_file(conn, "link_type_table_triggers.sql")
+
+
+def add_trigger_from_file(conn, qry_file: str) -> None:
+    qry_file = os.path.join(os.path.dirname(os.path.realpath(__file__)), "network/database_triggers", qry_file)
+    curr = conn.cursor()
+    sql_file = open(qry_file, "r")
+    query_list = sql_file.read()
+    sql_file.close()
+
+    # Run one query/command at a time
+    for cmd in query_list.split("#"):
+        try:
+            curr.execute(cmd)
+        except Exception as e:
+            msg = f"Error creating trigger: {e.args}"
+            logger.error(msg)
+            logger.info(cmd)
+    conn.commit()
```

Comment:
so hacky - i love it

I've written the same method before... except I split on ;

Reply:
I can't split on ';' because the triggers have ';' inside them....